### DISCUSSION TITLE
Vevor MPPT Solar Charge Controller

### DISCUSSION THREAD

**Discussion by joba-1** [arendst/Tasmota#16583] [Q&A] [answered]
Hi,

I got a Vevor 40A MPPT solar charge controller with an RS485 port.
The seller says it talks modbus and I have the physical spec (Volts, Pins). 

I don't have experience with modbus. Given I have the RS485 hardware interface, is that already enough info to talk to the device? 

I'd like to query the usual stuff PV voltage, Amps, battery voltage and amps, ...
I also need a way to set user settings for the device to support LiFePo charging.

According to [Modbus Spec V 1.02](https://modbus.org/docs/Modbus_over_serial_line_V1_02.pdf) chapter 4.2. every device needs to document specific stuff, so I guess I need more. Right?

**Comment by altelch** [comment, marked as answer]
Hi,

the Vevor 40A MPPT is a eSmart3 AFAIK:

https://skagmo.com/page.php?p=documents/04_esmart3_review

This device doesn't speak modbus but a proprietary protocol described here: https://github.com/skagmo/esmart_mppt/files/3401128/esmart3-serial-comm.pdf

Greetings,

       Heiko.

**Reply by joba-1** [reply]
Hi Heiko,

you are right, the pictures match. Thank you!

The review talks about a firmware problem: MPPT drops to 45W and is stuck until PV is disconnected 2-3 times a day.
Looks like pretty useless then. Do you know if this bug shows up on every device?

Or could you recommend a similar device? Minimum requirement would be supporting one panel with max ~300W / 35V, charging a 4S-LiFePo, with known protocol to query (and set, if needed) values - ideally with tasmota support.

Gruß, Joachim

**Reply by altelch** [reply]
Hi,

this "failure" seems to be related to the battery voltage. If your PV voltage is close to your battery voltage then the Firmware sometimes only found a local maximum and got stuck there. Don't know if this problem still exists. There are new firmware releases since then.

The epever tracer series is often recomended and has a Modbus interface.

Greetings,

            Heiko.

**Reply by altelch** [reply]
Hi,

falls interessiert könnten wir über eine tmp-mail reale Kontaktdaten austauschen. Habe auch einen eSmart3 am Start...

Gruß,
            Heiko.

**Reply by joba-1** [reply]
Gern. Aber der Aufwand mit tmp-mail ist glaube ich nicht nötig. Meine Adresse ist sowieso in jeder Menge Git Logs meiner [Repos](https://github.com/joba-1?tab=repositories) zu finden :)

**Comment by joba-1** [comment]
...and now there is an [eSmart3 library](https://github.com/joba-1/Joba_ESmart3) where most documented RS485 commands of the device have convenient methods - including toggling load, limiting current, setting battery parameters and monitoring status.

Maybe it helps someone to write a tasmota energy device ;)

**Reply by barbudor** [reply]
Better that someone who own the device can do it

**Comment by adekloet** [comment]
I didin't want to revive this old topic, but I have a li time 30A MPPT charger with an RJ12 RS4385 port connected to a waveshare RS485 to ETH connector.  I managed to get the pin layout from a Li-Time engineer, but still I am not able to get data even with diffrent speed tries etc. anybody any Idea what to try? 

![image](https://github.com/user-attachments/assets/6c35acec-8423-4020-96e0-81d05907dbc3)

This is the pin-out I got from Li-time.

thanks,

I have the waveshare integrated as Modbus in home assistant.

**Reply by joba-1** [reply]
usually devices on the rs485 bus only send data if you request it. Ask your engineer for the proper commands i.e the protocol the charger speaks, something similar to the pdf for the esmart.
